sync/sync.go

```diff
@@ -143,20 +140,24 @@ func (s *Synchronizer) syncBlocks(syncCtx context.Context) {
 
 	for {
 		select {
-		case <-errChan:
-			streamCancel() // cancel all running tasks
-			streamCtx, streamCancel = context.WithCancel(syncCtx)
-			nextHeight = s.nextHeight()
-			s.log.Warnw("Rolling back sync process", "height", nextHeight)
-		case <-syncCtx.Done():
-			fetchers.Wait()
-			verifiers.Wait()
-			return
+		case <-streamCtx.Done():
+			select {
+			case <-syncCtx.Done():
+				streamCancel()
+				fetchers.Wait()
+				verifiers.Wait()
+				return
+			default:
+				streamCtx, streamCancel = context.WithCancel(syncCtx)
+				nextHeight = s.nextHeight()
+				s.log.Warnw("Rolling back sync process", "height", nextHeight)
+			}
 		default:
 			curHeight := nextHeight
 			curStreamCtx := streamCtx
+			curCancel := streamCancel
 			fetchers.Go(func() stream.Callback {
-				return s.fetcherTask(curStreamCtx, curHeight, verifiers, errChan)
+				return s.fetcherTask(curStreamCtx, curHeight, verifiers, curCancel)
```

Comment:
Please update the code as follows:

```go
fetchers.Go(func() stream.Callback {
	return s.fetcherTask(streamCtx, nextHeight, verifiers, streamCancel)
})
nextHeight++
```

Reply:
That would be incorrect tho, no?

`streamCtx`, `nextHeight`, `streamCancel` can all change before goroutine starts execution.

Reply:
Yes, you are right. I forgot it would be started in a go routine. In that case, please assign all of them in a single line as follow:

```go
curHeight, curStreamCtx, curCancel := nextHeight, streamCtx, streamCancel
```

Reply:
Neat, done